cirq-core/cirq/ops/boolean_hamiltonian.py

```diff
@@ -0,0 +1,421 @@
+
+    CNOT(i, j) @ CNOT(j, k) == CNOT(j, k) @ CNOT(i, k) @ CNOT(i, j)
+    ───@───────       ───────@───@───
+       │                     │   │
+    ───X───@───   =   ───@───┼───X───
+           │             │   │
+    ───────X───       ───X───X───────
+
+    Args:
+        cnots: A list of CNOTS, encoded as integer tuples (control, target).
+        flip_control_and_target: Whether to flip control and target.
+
+    Returns:
+        A Boolean that tells whether a simplification has been performed.
+        The CNOT list, potentially simplified.
+    """
+    target, control = (0, 1) if flip_control_and_target else (1, 0)
+
+    # We investigate potential pivots sequentially.
+    for j in range(1, len(cnots) - 1):
+        # First, we look back for as long as the targets (resp. controls) are the same.
+        # They all commute, so all are potential candidates for being simplified.
+        common_a: Dict[int, int] = {}
+        for i in range(j - 1, -1, -1):
+            if cnots[i][control] != cnots[j][control]:
+                break
+            # We take a note of the control (resp. target).
+            common_a[cnots[i][target]] = i
+
+        # Next, we look forward for as long as the controls (resp. targets) are the
+        # same. They all commute, so all are potential candidates for being simplified.
+        common_b: Dict[int, int] = {}
+        for k in range(j + 1, len(cnots)):
+            if cnots[j][target] != cnots[k][target]:
+                break
+            # We take a note of the target (resp. control).
+            common_b[cnots[k][control]] = k
+
+        # Among all the candidates, find if they have a match.
+        keys = common_a.keys() & common_b.keys()
+        for key in keys:
+            assert common_a[key] != common_b[key]
+            # We perform the swap which removes the pivot.
+            new_idx: List[int] = (
+                [idx for idx in range(0, j) if idx != common_a[key]]
+                + [common_b[key], common_a[key]]
+                + [idx for idx in range(j + 1, len(cnots)) if idx != common_b[key]]
+            )
+            # Since we removed the pivot, the length should be one fewer.
+            assert len(new_idx) == len(cnots) - 1
+            cnots = [cnots[idx] for idx in new_idx]
+            return True, cnots
+
+    return False, cnots
+
+
+def _simplify_cnots(cnots: List[Tuple[int, int]]) -> List[Tuple[int, int]]:
+    """Takes a series of CNOTs and tries to applies rule to cancel out gates.
+
+    Args:
+        cnots: A list of CNOTs represented as tuples of integer (control, target).
+
+    Returns:
+        A Boolean saying whether a simplification has been found.
+        The simplified list of CNOTs.
+    """
+
+    found_simplification = True
+    while found_simplification:
+        for simplify_fn, flip_control_and_target in itertools.product(
+            [_simplify_commuting_cnots, _simplify_cnots_triplets], [False, True]
```

Comment:
again, I'm not convinced that this is the most efficient way of simplifications.